drivers/Watchdog.cpp

```diff
@@ -20,46 +20,89 @@
 
 namespace mbed {
 
-watchdog_status_t Watchdog::start(const uint32_t timeout)
-{
-    if (timeout == 0) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
-    }
+Watchdog *Watchdog::_first = NULL;
 
-    if (timeout > max_timeout()) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
-    }
+Watchdog::Watchdog(uint32_t timeout, const char *const str): _name(str)
+{
+    _current_count = 0;
+    _is_initialized = false;
+    _next = NULL;
+    _max_timeout = timeout;
+}
 
-    watchdog_config_t config;
-    config.timeout_ms = timeout;
+void Watchdog::add_to_list()
+{
+    this->_next = _first;
+    _first = this;
+    _is_initialized =  true;
+}
 
-    return hal_watchdog_init(&config);
+void Watchdog::start()
+{
+    MBED_ASSERT(!_is_initialized);
+    core_util_critical_section_enter();
+    add_to_list();
+    core_util_critical_section_exit();
 }
 
 
 void Watchdog::kick()
 {
-    hal_watchdog_kick();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    _current_count = 0;
+    core_util_critical_section_exit();
 }
 
-
-watchdog_status_t Watchdog::stop()
+void Watchdog::stop()
 {
-    return hal_watchdog_stop();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    remove_from_list();
+    core_util_critical_section_exit();
 }
 
-
-uint32_t Watchdog::reload_value() const
+void Watchdog::remove_from_list()
 {
-    return hal_watchdog_get_reload_value();
+    Watchdog *cur_ptr = _first,
+              *prev_ptr = NULL;
+    while (cur_ptr != NULL) {
+        if (cur_ptr == this) {
+            if (cur_ptr == _first) {
+                prev_ptr = _first;
```

Comment:
`_first` is `this`, can we work with `this` instead?

Reply:
No,We just make sure cur_ptr aginst _first node,but tested this code robust enough.

Reply:
I would agree with @donatieng in the sense that `cur_ptr` and `_first` are variable in this context.
When reading the code, it might be easier to replace usage of `cur_ptr` with `this` in this block.
```c++
// look up for this watchdog in the list to find its predecessor
while ((cur_ptr != this) && (cur_ptr != NULL)) {
    prev_ptr = cur_ptr;
    cur_ptr = cur_ptr->next;
}
if (cur_ptr == NULL) {
    // this was not in the list
    return;
}

if (prev_ptr == NULL) {
    // this was first, set _next as the new _first.
    _first = this->_next;
} else {
    // attach previous watchdog to our next
    prev_ptr->_next = this->_next;
}
// detach us from _next
this->_next = NULL;
_is_initialized = false;
```